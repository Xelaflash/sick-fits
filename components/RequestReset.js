import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Form from './styles/Form';
import useForm from '../lib/useForm';
import DisplayError from './ErrorMessage';

const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      code
      message
    }
  }
`;

export default function RequestReset() {
  const { inputs, handleChange } = useForm({
    email: '',
  });

  const [requestReset, { data, loading, error }] = useMutation(
    REQUEST_RESET_MUTATION,
    {
      variables: inputs,
    }
  );

  async function handleSubmit(e) {
    e.preventDefault();
    // send email and password to graphQl API
    await requestReset().catch(console.error);
    // console.log(res);
    // console.log({ data, loading, error });
  }

  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Reset Password</h2>
      <DisplayError error={error} />
      <fieldset
        className="inputs-subgrid"
        disabled={loading}
        aria-busy={loading}
      >
        {data?.sendUserPasswordResetLink === null && (
          <p>Reset Link sent. Check your email</p>
        )}
        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            placeholder="x@x.com"
            autoComplete="email"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>

        <button type="submit">Reset</button>
      </fieldset>
    </Form>
  );
}

export { REQUEST_RESET_MUTATION };
