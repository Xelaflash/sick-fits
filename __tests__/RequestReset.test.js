import { render, screen } from '@testing-library/react';
import { MockedProvider } from '@apollo/react-testing';
import userEvent from '@testing-library/user-event';
import RequestReset, {
  REQUEST_RESET_MUTATION,
} from '../components/RequestReset';

const email = 'rien@rien.com';

const mock = [
  {
    request: {
      query: REQUEST_RESET_MUTATION,
      variables: { email },
    },
    result: {
      data: { sendUserPasswordResetLink: null },
    },
  },
];

describe('<RequestReset />', () => {
  it('1. renders and matches snapshot', () => {
    const { container } = render(
      <MockedProvider>
        <RequestReset />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });
  it('2. Calls mutation when submitted', async () => {
    const { container, debug } = render(
      <MockedProvider mocks={mock}>
        <RequestReset />
      </MockedProvider>
    );
    // we type into user input field
    userEvent.type(screen.getByPlaceholderText(/x@x.com/), email);
    userEvent.click(screen.getByText('Reset'));
    debug();
    const success = await screen.findByText(
      /Reset Link sent. Check your email/i
    );
    // screen.debug(success);
    expect(success).toBeInTheDocument();
  });
});
