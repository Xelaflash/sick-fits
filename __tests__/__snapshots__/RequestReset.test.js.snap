// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RequestReset /> 1. renders and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 iaUpjc"
    method="POST"
  >
    <h2>
      Reset Password
    </h2>
    <fieldset
      aria-busy="false"
      class="inputs-subgrid"
    >
      <label
        for="email"
      >
        Email
        <input
          autocomplete="email"
          name="email"
          placeholder="x@x.com"
          type="email"
          value=""
        />
      </label>
      <button
        type="submit"
      >
        Reset
      </button>
    </fieldset>
  </form>
</div>
`;
